com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -940,11 +940,25 @@ public void SetSingleton()
             OnSingletonReady?.Invoke();
         }
 
+        /// <summary>
+        /// Returns the root GameObject of the NetworkManager's GameObject
+        /// </summary>
+        internal GameObject GetRootParent(GameObject currentGameObject = null)
+        {
+            currentGameObject = currentGameObject == null ? gameObject : currentGameObject;
+            if (currentGameObject.transform.parent != null)
+            {
+                return GetRootParent(currentGameObject.transform.parent.gameObject);
+            }
+            return currentGameObject;
+        }
+
         private void OnEnable()
         {
             if (DontDestroy)
             {
-                DontDestroyOnLoad(gameObject);
+                // We must always move the root GameObject into the DDOL
+                DontDestroyOnLoad(GetRootParent());
```

Comment:
I have mixed feelings about this.
We might just throw an exception here (similar to `DontDestroyOnLoad` API throwing "only root objects can be marked DDOL") and expect user to do the right thing so that they could be much more intentional.
One thing I don't want to encourage here is to fix issues "automagically" for the user — they should face invalid use cases and align things accordingly.

Reply:
so to be clear, maybe the fix is just throwing a descriptive exception here.
having said that, it looks like issue #1417 never got an exception but instead got a "not found" error.
perhaps we should properly throw an exception to better inform user what's going on so that they could understand and react?

Reply:
> **maybe the fix is just throwing a descriptive exception here.**
[Object.DontDestroyOnLoad](https://docs.unity3d.com/ScriptReference/Object.DontDestroyOnLoad.html) is basically a "one off" method from [SceneManager.MoveGameObjectToScene](https://docs.unity3d.com/ScriptReference/SceneManagement.SceneManager.MoveGameObjectToScene.html) which has the following requirement:
> You can only move root GameObjects from one Scene to another. This means the GameObject to move must not be a child of any other GameObject in its Scene.

This makes it a requirement to provide the root GameObject to Object.DontDestroyOnLoad.  

> **never got an exception but instead got a "not found" error.**
Don't forget the second part of what the user reports in the issue:
> and everything related to the network will not working. 

It can get many different types of errors when this happens for different reasons.

So, keeping in alignment with the default UnityEngine requirements solves the issue.  For NetworkManager (only), it should be perfectly valid to place the NetworkManager component's GameObject as a child under another GameObject (or several depending upon what the user need might be).

Throwing an exception I am going to try to stay away from for this one.

Reply:
I disagree. DDOL on a child object is not OK in Unity therefore I believe we should not magically fix user errors here but instead just shout them out loud.
We could chat about this but I'm also happy to defer this one to @andrews-unity or @mattwalsh-unity

Reply:
Well, I do agree about the requirement to send the root parent into the DDOL.  I also think that we could throw an exception or warn the user in the editor.  The only thing I wonder about is whether it is the right thing to throw an exception under this condition since everything NGO is really "a child" (component) of GameObject.  If you have GameObject-A as a root parent with no NGO components and GameObject-B with a NetworkObject component as the child of GameObject-A but you want to make sure the NetworkObject (GameObject-B) itself is moved into the DDOL then the big question is do we assert/except on this condition or do we just hunt for the root parent and move that into DDOL?

I am open to both lines of thinking and will see if I can stir up a discussion on this topic to discuss further.

Reply:
Ok, per our slack exchange, I'd like to go ahead and do the following:

- (having conferred with @TwoTenPvP) Make NetworkManager always "DDOL" - remove the setting.  That is we never ever want to have the act of changing a scene implicitly tear down a NM
- require NetworkManager to be sitting on a GameObject which itself is a root-level component, else complain as loudly and proudly and early as possible (ideally in the editor, ideally to block running at all)
- (not in this PR) investigate how we can make "NM is a component of a GO" not a thing